### DISCUSSION TITLE
How to set PWM based on a variable form a JSON formatted WebQuery GET?

### DISCUSSION THREAD

**Discussion by danpadure** [arendst/Tasmota#17664] [Q&A] [answered]
I have a very simple JSON webpage:

{"id":"sensor-duty-cycle","value":100,"state":"100 %"}

I need to extract the "sensor-duty-cycle" variable  and assign it to a PWM output.  I see that if I configure an output as PWM, then it's named "Dimmer"

I don't understand how to extract the variable from the PWM 
I can run the command, but how to extract the variable?
21:44:26.562 CMD: WebQuery http://192.168.12.80/sensor/duty-cycle GET
21:44:26.715 RSL: RESULT = {"WebQuery":"Done"}

Thanks

**Comment by barbudor** [comment, marked as answer]
You need a custom build with #USE_WEBSEND_RESPONSE
This will allow the response from your server to be pushed to the Rules engine in the form ` {"WebQuery":{"id":"sensor-duty-cycle","value":100,"state":"100 %"}}` 
Then a rule such as `on WebQuery#id do PWM %value% endon` should work

Without USE_WEBSEND_RESPONSE, it will not work

**Reply by danpadure** [reply]
Thanks, it worked

The actual rule is like this:

Rule1
  ON Time#Minute DO backlog WebQuery http://192.168.12.80/sensor/duty-cycle/ GET ENDON
  on WebQuery#value do DIMMER %value% endon

**Reply by barbudor** [reply]
Indeed I miss read the proper name of the key ;)

**Reply by sfromis** [reply]
If you want the specific PWM value, without the dimmer handling, you can also configure for the PWM command to be the one to use.

**Reply by danpadure** [reply]
> If you want the specific PWM value, without the dimmer handling, you can also configure for the PWM command to be the one to use.

How?  I found no reference to any PWM command

**Reply by danpadure** [reply]
I also found a trick to run it every 10 seconds.  
Bu using ruletimer 10 that gets reset at every run

**Reply by sfromis** [reply]
[PWM command](https://tasmota.github.io/docs/Commands/#PWM). You did look for a command?

**Comment by danpadure** [comment]
Obviously I did NOT search well enough
I used the search on top of the page of the main docs

**Reply by sfromis** [reply]
That search can be wonky.

**Reply by danpadure** [reply]
@sfromis do you know what's the difference between PWM and PWMi, and how does SetOption15 affect this?
apparently if you SetOption15 1 then dimmer command works.
But what else is different?
Docs only says "GPIOs are used in two modes depending on SetOption15: either as lights or as pure PWM" but it does not explain the difference
Using the PWM with dimmer (0-100) commands is any different than using pure PWM commands?

**Reply by sfromis** [reply]
`PWMi` is inverted, otherwise the same. Not related to SetOption15 used to enable basic PWM mode instead of dimmer control. What I wrote was that **if** you want specific PWM control (of duty cycle), instead of the more coarse-grained dimmer control, **then** you may want the basic mode, and control via the `PWM` command, which allows fine-grained and precise control. This is the difference.

**Reply by danpadure** [reply]
ok, PWMi I understand.
However although I do know what duty cycle, PWM frequency are, I don't know the more subtle differences that you mention
But let's keep it simple: I need PWM to modulate a steering wheel heating resistor, to be able to change it just like you would from a normal PWM voltage regulator
So my guess is that frequency does not matter, and the subtle differences should not matter neither (or "either", how is it grammatically correct?)
So should I use dimmer command that also gives me a nice UI, or should I use PWM command, or something else?

**Reply by barbudor** [reply]
DIMMER commadn is part of Tasmota Light model which has some specifics:
- range is 0..100 (that's why sfromis say it's coarse setting)
- a gamma correction curve can be applied (based on some setoptions)
- ther is a relation between POWER and DIMMER so that if you set DIMMER 0, that will also set POWER OFF. If you set DIMMER >0, that will imply POWER ON

PWM is just setting a PWM output
- range is 0..1023 so this is a finer grain
- no relation with POWER

Up to you

**Reply by danpadure** [reply]
ok, I'll disable gamma correction, otherwise I'm fine with small grain, and I like that power goes to OFF when DIMMER=0
so I guess DIMMER is quite useful for me in this case :)
